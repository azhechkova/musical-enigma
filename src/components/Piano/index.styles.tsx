import styled from 'styled-components';

const BasicPiano = styled.div`
  display: flex;
  border-radius: 4px;
`;

export default BasicPiano;
