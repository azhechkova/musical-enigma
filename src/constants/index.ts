export const noteList: NoteType[] = [
  {
    key: 'KeyA',
    keyName: 'A',
    note: 'C4',
    type: 'tone',
  },
  {
    key: 'KeyW',
    keyName: 'W ',
    note: 'C#4',
    type: 'semitone',
  },
  {
    key: 'KeyS',
    keyName: 'S',
    note: 'D4',
    type: 'tone',
  },
  {
    key: 'KeyE',
    keyName: 'E',
    note: 'D#4',
    type: 'semitone',
  },
  {
    key: 'KeyD',
    keyName: 'D',
    note: 'E4',
    type: 'tone',
  },
  {
    key: 'KeyF',
    keyName: 'F',
    note: 'F4',
    type: 'tone',
  },
  {
    key: 'KeyT',
    keyName: 'T',
    note: 'F#4',
    type: 'semitone',
  },
  {
    key: 'KeyG',
    keyName: 'G',
    note: 'G4',
    type: 'tone',
  },
  {
    key: 'KeyY',
    keyName: 'Y',
    note: 'G#4',
    type: 'semitone',
  },
  {
    key: 'KeyH',
    keyName: 'H',
    note: 'A4',
    type: 'tone',
  },
  {
    key: 'KeyU',
    keyName: 'U',
    note: 'A#4',
    type: 'semitone',
  },
  {
    key: 'KeyJ',
    keyName: 'J',
    note: 'B4',
    type: 'tone',
  },
  {
    key: 'KeyK',
    keyName: 'K',
    note: 'C5',
    type: 'tone',
  },
  {
    key: 'KeyO',
    keyName: 'O',
    note: 'C#5',
    type: 'semitone',
  },
  {
    key: 'KeyL',
    keyName: 'L',
    note: 'D5',
    type: 'tone',
  },
  {
    key: 'KeyP',
    keyName: 'P',
    note: 'D#5',
    type: 'semitone',
  },
  {
    keyName: 'Semicolon',
    key: 'Semicolon',
    note: 'E5',
    type: 'tone',
  },
  { key: 'Quote', note: 'F5', type: 'tone' },
  {
    keyName: 'Bracket',
    key: 'BracketRight',
    note: 'F#5',
    type: 'semitone',
  },
  {
    keyName: 'Backslash',
    key: 'Backslash',
    note: 'G5',
    type: 'tone',
  },
];

const getKeysList = (notes: NoteType[]): KeysType =>
  notes.reduce<KeysType>((prev, curr) => {
    prev[curr.key] = curr;

    return prev;
  }, {});

export const keysList = getKeysList(noteList) as KeysType;
