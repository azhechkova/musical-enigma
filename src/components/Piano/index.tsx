import React, { useCallback, useEffect, useRef, useState } from 'react';

import * as Tone from 'tone';
import BasicPiano from './index.styles';
import { keysList } from '../../constants';
import Note from '../Note';

interface PianoProps {
  notes: NoteType[];
}
type PressedKeyType = Record<string, boolean>;

const Piano = ({ notes }: PianoProps): JSX.Element => {
  const synth = useRef<Tone.Synth | null>(null);
  const [pressedKeys, setPressedKeys] = useState<PressedKeyType>({});

  const setupSynth = (): void => {
    synth.current = new Tone.Synth().toDestination();
  };

  useEffect(() => {
    const formatKeys = notes?.reduce<PressedKeyType>((prev, next) => {
      prev[next.key] = false;

      return prev;
    }, {});

    setPressedKeys(formatKeys);
  }, [notes]);

  const playNote = useCallback((note: string): void => {
    if (!synth.current) setupSynth();
    synth.current?.triggerAttack(note, '8n');
  }, []);

  const onKeyPress = useCallback(
    (e: KeyboardEvent): void => {
      const note = keysList[e.code];

      if (note?.name) {
        setPressedKeys({ ...pressedKeys, [e.code]: true });
        playNote(note.name);
      }
    },
    [playNote, pressedKeys],
  );

  const stopPlay = useCallback((): void => {
    synth.current?.triggerRelease(Tone.now());
  }, []);

  const onKeyUp = useCallback(
    (e: KeyboardEvent): void => {
      const note = keysList[e.code];

      if (note?.name) {
        setPressedKeys({ ...pressedKeys, [e.code]: false });

        stopPlay();
      }
    },
    [pressedKeys, stopPlay],
  );

  useEffect(() => {
    window.addEventListener('keydown', onKeyPress);
    window.addEventListener('keyup', onKeyUp);

    return () => {
      window.removeEventListener('keydown', onKeyPress);
      window.removeEventListener('keyup', onKeyUp);
    };
  }, [onKeyPress, onKeyUp]);

  return (
    <BasicPiano>
      {notes.map(note => (
        <Note
          stopPlay={stopPlay}
          key={note.key}
          note={{ ...note, isPressed: pressedKeys[note.key] }}
          playNote={playNote}
        />
      ))}
    </BasicPiano>
  );
};

export default Piano;
