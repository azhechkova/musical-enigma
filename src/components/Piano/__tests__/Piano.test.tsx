import React from 'react';
import { render } from '@testing-library/react';

import Piano from '..';
import { noteList } from '../../../constants';

describe('Piano', () => {
  it('renders properly', () => {
    const { container: piano } = render(<Piano notes={noteList} />);

    expect(piano).toMatchSnapshot();
  });
});
