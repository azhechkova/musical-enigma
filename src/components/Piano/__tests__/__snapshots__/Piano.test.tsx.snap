// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Piano renders properly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 4px;
}

.c1 {
  display: block;
  padding: 0;
  border-color: var(--white-hover);
  cursor: pointer;
  height: 200px;
  width: 50px;
  box-shadow: -1px 0 0 var(--white-hover) inset,0 0 5px hsl(0,0%,80%) inset, 0 0 3px var(--black);
  background: linear-gradient(to bottom,hsl(0,0%,93%) 0%,var(--white) 100%);
}

.c1:first-child {
  border-radius: 4px 0 4px 4px;
}

.c1:last-child {
  border-radius: 0 4px 4px 4px;
}

.c1:active {
  border-top: 1px solid #777;
  border-left: 1px solid #999;
  border-bottom: 1px solid #999;
  box-shadow: 2px 0 3px rgba(0,0,0,0.1) inset, -5px 5px 20px rgba(0,0,0,0.2) inset,0 0 3px rgba(0,0,0,0.2);
  background: linear-gradient(to bottom,var(--white) 0%,#e9e9e9 100%);
}

.c2 {
  width: 2em;
  margin: 0 -1em;
  z-index: 10;
  position: relative;
  height: 100px;
  box-shadow: -1px -1px 2px var(--black-hover) inset, 0 -5px 2px 3px var(--black) inset,0 2px 4px var(--black-hover);
  background: linear-gradient(45deg,hsl(0,0%,13%) 0%,hsl(0,0%,33%) 100%);
}

.c2:active {
  box-shadow: -1px -1px 2px rgba(232,237,223,0.2) inset, 0 -2px 2px 3px rgba(36,36,35,0.6) inset, 0 1px 2px rgba(36,36,35,0.5);
  background: linear-gradient(to right,#444 0%,#222 100%);
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
      title="C4"
      type="button"
    />
    <button
      class="c1 c2"
      title="C#4"
      type="button"
    />
    <button
      class="c1"
      title="D4"
      type="button"
    />
    <button
      class="c1 c2"
      title="D#4"
      type="button"
    />
    <button
      class="c1"
      title="E4"
      type="button"
    />
    <button
      class="c1"
      title="F4"
      type="button"
    />
    <button
      class="c1 c2"
      title="F#4"
      type="button"
    />
    <button
      class="c1"
      title="G4"
      type="button"
    />
    <button
      class="c1 c2"
      title="G#4"
      type="button"
    />
    <button
      class="c1"
      title="A4"
      type="button"
    />
    <button
      class="c1 c2"
      title="A#4"
      type="button"
    />
    <button
      class="c1"
      title="B4"
      type="button"
    />
    <button
      class="c1"
      title="C5"
      type="button"
    />
    <button
      class="c1 c2"
      title="C#5"
      type="button"
    />
    <button
      class="c1"
      title="D5"
      type="button"
    />
    <button
      class="c1 c2"
      title="D#5"
      type="button"
    />
    <button
      class="c1"
      title="E5"
      type="button"
    />
    <button
      class="c1"
      title="F5"
      type="button"
    />
    <button
      class="c1 c2"
      title="F#5"
      type="button"
    />
    <button
      class="c1"
      title="G5"
      type="button"
    />
  </div>
</div>
`;
